### PR TITLE
Grunt

### PR DESCRIPTION
**Pull request by johnnymo87**
### Ready for pull!

I want to start using the same tool chain that all other angular devs are using. To that end, I'm following [this guide](http://www.yearofmoo.com/2013/09/advanced-testing-and-debugging-in-angularjs.html), liberally copying what I see in [the video](https://www.youtube.com/watch?v=AKwqfHm-3ZQ) and [the seed app](https://github.com/yearofmoo/angularjs-seed-repo). 

Follow the readme on the seed app to get setup with this toolchain.

I've moved the jasmine unit tests to their new location, but they're blowing up because they depend on `jasmine-jquery`, a dependency we should lose in exchange for the angular way of writing jasmine.

### CODE REVIEW

File: test/unit/services/editors_factory_spec.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  beforeEach(inject(function($injector) {
+    factory = $injector.get("EditorsFactory");
+  }));
+
+  it('jasmine works', function(){
+    var color = 'green';
+    expect(color).toBe('green')
+  });
+
+  describe('jQuery-style jasmine', function() {
```
Comment (johnnymo87) [inline test/unit/services/editors_factory_spec.js]: Tests below this line fail because they are depending on `jasmine-jquery`.

File: bower.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "yearofmoo-angularjs-seed-repo",
```
Comment (yggie) [inline bower.json]: this needs to be updated

File: package.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (yggie) [inline package.json:1]: do we still need `package.json` since we have `bower.json`?

File: .gitignore [hunk 4]
```diff
@@ -43,3 +43,10 @@ public/assets/**
 
 # rvm
 .rvmrc
+
+# angular-seed-repo
+.agignore
```
Comment (yggie) [inline .gitignore:48]: what is in the `.agignore` file?

File: package.json (hunk 3 above)
Comment (johnnymo87) [inline package.json:1]: No clue, I'm just monkey-see-monkey-do-ing this. But notice that there's two commands -- `npm install` and later `bower install`.

File: .gitignore (hunk 4 above)
Comment (johnnymo87) [inline .gitignore:48]: It's got something to do with [Ag](https://github.com/ggreer/the_silver_searcher), a tool we both have in `yadr`.

Comment (yggie) [inline .gitignore:48]: +1000 discovering `.agignore` just made my life 10x easier

**Comment by yggie** [comment]
@johnnymo87 can you try adding `npm install` under `before_install` and replace `bundle exec rake jasmine:ci` with `karma start --single-run` or any grunt equivalent in `.travis.yml`. If that works out, we could completely remove `jasmine` related gems from the `Gemfile`

File: .gitignore [hunk 5]
```diff
@@ -43,3 +43,10 @@ public/assets/**
 
 # rvm
 .rvmrc
+
+# angular-seed-repo
```
Comment (yggie) [inline .gitignore]: this could be removed, looks like it was pulled in when you were following that blog

File: app/assets/javascripts/services/editors_factory.js [hunk 6]
```diff
@@ -27,7 +27,7 @@ CodealiaApp.factory("EditorsFactory", function() {
   }
 
   self.setEditor = function(scope, element, attrs) {
-    self.editor = ace.edit(element.find("#passions-editor")[0]);
+    self.editor = ace.edit(element.children()[0]);
```
Comment (yggie) [inline app/assets/javascripts/services/editors_factory.js]: could we use `querySelector` as a replacement for `find`?

**Comment by yggie** [comment]
@PeteBoucher this has been reviewed by me, and I am happy to merge this in. I want you to have a look at it too, since this will have an impact on your part as well.

File: .jshintrc [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    "noarg": true,
+    "regexp": true,
+    "undef": true,
+    "unused": true,
+    "trailing": true,
+    "smarttabs": true,
+    "globals": {
+        "angular": false
+    }
+}
```
Comment (PeteBoucher) [inline .jshintrc:21]: Where do jshints manifest? Is this a dev environment thing specific to vim? Could it be ignored?

File: app/assets/javascripts/services/editors_factory.js [hunk 8]
```diff
@@ -24,18 +24,18 @@ CodealiaApp.factory("EditorsFactory", function() {
     editor.setShowPrintMargin(false);
 
     editor.getSession().on("change", self.updatePreview);
-  }
+  };
 
   self.setEditor = function(scope, element, attrs) {
-    self.editor = ace.edit(element.find("#passions-editor")[0]);
+    self.editor = ace.edit(element[0].querySelector('#passions-editor'));
```
Comment (PeteBoucher) [inline app/assets/javascripts/services/editors_factory.js:30]: Why this change? I feel like it was more readable before.

Comment (johnnymo87) [inline app/assets/javascripts/services/editors_factory.js:30]: Because `.find()` is jQuery and not jQlite.

File: .jshintrc (hunk 7 above)
Comment (johnnymo87) [inline .jshintrc:21]: `jshint` is a grunt task that produces a report in your terminal. It's not specific to any editor. It's usually run as part of a 'full suite' test run (e.g. jshint, karma, protractor).

File: spec/javascripts/services/editors_factory_spec.js [hunk 9]
```diff
@@ -1,136 +0,0 @@
...
-        el.html("hello there");
-        factory.updateEditor();
-      });
-
-      it("should update the contents of the editor", function() {
-        expect(factory.editor.getValue()).toEqual("hello there");
-      });
-    });
-  });
-});
```
Comment (PeteBoucher) [inline spec/javascripts/services/editors_factory_spec.js:136]: Why the move from `spec/javascript` to `test/unit`? Do we really need a third top level directory for tests?

File: test/karma-unit.conf.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      'bower_components/angular-route/angular-route.js',
+      'bower_components/angular-mocks/angular-mocks.js',
+      'app/assets/javascripts/**/*.js',
+      'test/unit/**/*.js',
+      'vendor/assets/javascripts/ace/**/*.js'
+    ],
+    plugins: [
+      'karma-phantomjs-launcher',
+      'karma-chrome-launcher',
+      'karma-jasmine'
```
Comment (PeteBoucher) [inline test/karma-unit.conf.js:14]: We removed jasmine, right?

Comment (yggie) [inline test/karma-unit.conf.js:14]: we removed `jasmine` related gems, the `jasmine` framework is still needed for testing

File: spec/javascripts/services/editors_factory_spec.js (hunk 9 above)
Comment (yggie) [inline spec/javascripts/services/editors_factory_spec.js:136]: I feel it gives a clear separation between the rails and angular portion of the app. Also, having it under `spec/` can be misleading, as it suggests `rspec` can be used on `spec/javascript`

Comment (johnnymo87) [inline spec/javascripts/services/editors_factory_spec.js:136]: Usually an angular app has a top-level directory structured like this:

.
├── app
...
└── test
    ├── e2e
    │   └── scenarios.js
    ├── karma.conf.js
    ├── protractor-conf.js
    └── unit
     │   ├── controllersSpec.js
     │   ├── directivesSpec.js
     │   ├── filtersSpec.js
     │   └── servicesSpec.js

The `e2e` directory is a sibling to the `unit` directory. `conf` files for each are side-by-side too. We don't have to adopt this convention, but it makes sense if we plan to do e2e testing with protractor.

Comment (PeteBoucher) [inline spec/javascripts/services/editors_factory_spec.js:136]: I like the idea of convention over config. but there seems to be an overwhelming amount of config already in this Angular test setup, may aswel use it to reduce complexity in the app structure. And since we already chose spec/ over test/ in our rails setup it feels more confusing to have both: 
if spec/ == rspec then test/ == vanilla unit tests
How about `spec/grunt` as a compromise?

Comment (yggie) [inline spec/javascripts/services/editors_factory_spec.js:136]: well `spec/grunt` isn't better than `spec/javascripts`, so I would just stick with `spec/javascripts`. What do you think @johnnymo87? Could we stick with this for now?

Comment (johnnymo87) [inline spec/javascripts/services/editors_factory_spec.js:136]: I'm not opposed to it, but it will require that all the references for all the testing commands are updated in the `gruntfile`, `karma.conf`, and `protractor.conf`.

Comment (PeteBoucher) [inline spec/javascripts/services/editors_factory_spec.js:136]: If you want to pair on those changes @johnnymo87 let me know. If not I'm happy with whatever structure you decide is best.
